timemachine/fe/bar.py

```diff
@@ -99,70 +98,113 @@ def dG_dw(w):
     return dG_dw
 
 
-def bootstrap_bar(w_F, w_R, n_bootstrap=1000, timeout=10) -> Tuple[float, NDArray]:
-    """Subsample w_F, w_R with replacement and re-run BAR many times
+def ukln_to_ukn(u_kln: NDArray) -> Tuple[NDArray, NDArray]:
+    """Convert 2-state u_kln matrix to u_kn and N_k, i.e. the inputs expected by pymbar.MBAR.
+
+    NOTE: similar to https://pymbar.readthedocs.io/en/master/utils.html#pymbar.utils.kln_to_kn, but uses the (current)
+    timemachine convention where the first two axes correspond to evaluation state and sampling state respectively.
+
+    TODO: consider switching to pymbar convention?
 
     Parameters
     ----------
-    w_F : array
-        forward works
-    w_R : array
-        reverse works
+    u_kln : array (2, 2, N)
+        2-state u_kln matrix, where
+        * the first dimension (k) indexes the state for which we evaluate the energy
+        * the second dimension (l) indexes the state from which the configuration was sampled
+    """
+    k, l, n = u_kln.shape
+    assert k == l == 2
+    u_kn = u_kln.reshape(k, -1)
+    assert u_kn.shape == (k, l * n)
+    N_k = n * np.ones(l)
+    return u_kn, N_k
+
+
+DEFAULT_RELATIVE_TOLERANCE = 1e-6  # pymbar default 1e-7
+DEFAULT_MAXIMUM_ITERATIONS = 1_000  # pymbar default 10_000
+
+
+def df_and_err_from_u_kln(u_kln: NDArray, maximum_iterations: int = DEFAULT_MAXIMUM_ITERATIONS) -> Tuple[float, float]:
+    """Compute free energy difference and uncertainty given a 2-state u_kln matrix."""
+    u_kn, N_k = ukln_to_ukn(u_kln)
+    mbar = pymbar.MBAR(u_kn, N_k, maximum_iterations=maximum_iterations)
+    try:
+        df, ddf = mbar.getFreeEnergyDifferences()
+        return df[0, 1], ddf[0, 1]
+    except pymbar.utils.ParameterError:
+        # As of pymbar 3.1.0, computation of the covariance matrix can raise an exception on incomplete convergence.
+        # In this case, return the unconverged estimate with NaN as uncertainty.
+        df = mbar.getFreeEnergyDifferences(compute_uncertainty=False)[0]
+        return df[0, 1], np.nan
+
+
+def df_from_u_kln(
+    u_kln: NDArray, initial_f_k: Optional[NDArray] = None, maximum_iterations: int = DEFAULT_MAXIMUM_ITERATIONS
+) -> float:
+    """Compute free energy difference given a 2-state u_kln matrix."""
+    u_kn, N_k = ukln_to_ukn(u_kln)
+    mbar = pymbar.MBAR(u_kn, N_k, initial_f_k=initial_f_k, maximum_iterations=maximum_iterations)
+    df = mbar.getFreeEnergyDifferences(compute_uncertainty=False)[0]
+    return df[0, 1]
+
+
+def bootstrap_bar(
+    u_kln: NDArray, n_bootstrap: int = 100, maximum_iterations: int = DEFAULT_MAXIMUM_ITERATIONS
+) -> Tuple[float, NDArray]:
+    """Given a 2-state u_kln matrix, subsample u_kln with replacement and re-run df_from_u_kln many times
+
+    Parameters
+    ----------
+    u_kln : array
+        2-state u_kln matrix
     n_bootstrap : int
-        # bootstrap samples
-    timeout : int
-        in seconds
+        number o bootstrap samples
```

Comment:
typo: o -> of

Reply:
Fixed in https://github.com/proteneer/timemachine/pull/1098/commits/6950d82b1d3b12b8fb1951bdd6020c009b826caa